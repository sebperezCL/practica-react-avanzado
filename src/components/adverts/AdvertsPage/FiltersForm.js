import React from 'react';
import T from 'prop-types';
import { Button, Input, Slider, Radio, Row, Col } from 'antd';

import TagsSelect from '../TagsSelect';
import FormField from '../../shared/FormField';
import { saleOptions, MIN_PRICE, MAX_PRICE } from '../definitions';
import styles from './FiltersForm.module.css';

export const defaultFilters = {
  name: '',
  sale: saleOptions.all.value,
  price: [],
  tags: [],
};

class FiltersForm extends React.Component {
  state = {
    ...this.props.initialFilters,
  };

  handleNameChange = ev => this.setState({ name: ev.target.value });
  handlePriceChange = price => {
    const [min, max] = price;
    if (min === MIN_PRICE && max === MAX_PRICE) {
      return this.setState({ price: [] });
    }
    this.setState({ price });
  };
  handleSaleChange = ev => this.setState({ sale: ev.target.value });
  handleTagsChange = tags => this.setState({ tags });

  handleSubmit = ev => {
    const { onSubmit } = this.props;
    ev.preventDefault();
    onSubmit(this.state);
  };

  render() {
    const { name, price, tags, sale } = this.state;
    const priceValue = price.length === 0 ? [MIN_PRICE, MAX_PRICE] : price;

    return (
      <form onSubmit={this.handleSubmit}>
        <Row className={styles.form}>
          <Col span={11}>
            <FormField label="By name">
              <Input
                placeholder="Name"
                onChange={this.handleNameChange}
                value={name}
              />
            </FormField>
            <FormField
              label={
                <>
                  By price
                  <strong style={{ margin: '0 5px' }}>
                    {priceValue.join(' - ')}
                  </strong>
                </>
              }
            >
              <Slider
                range
                defaultValue={priceValue}
                min={MIN_PRICE}
                max={MAX_PRICE}
                onChange={this.handlePriceChange}
              />
            </FormField>
          </Col>
          <Col span={11} offset={2}>
            <FormField label="By tags">
              <TagsSelect onChange={this.handleTagsChange} value={tags} />
            </FormField>
            <FormField label="By type">
              <Radio.Group
                options={Object.values(saleOptions)}
                onChange={this.handleSaleChange}
                value={sale}
              />
            </FormField>
          </Col>
          <Col span={24}>
            <Button
              className={styles.button}
              type="primary"
              htmlType="submit"
              block
            >
              Search
            </Button>
          </Col>
        </Row>
      </form>
    );
  }
}

FiltersForm.propTypes = {
  initialFilters: T.shape({
    name: T.string,
    sale: T.oneOf(Object.keys(saleOptions)),
    price: T.arrayOf(T.number),
    tags: T.arrayOf(T.string),
  }),
  onSubmit: T.func.isRequired,
};

FiltersForm.defaultProps = {
  initialFilters: defaultFilters,
};

export default FiltersForm;
