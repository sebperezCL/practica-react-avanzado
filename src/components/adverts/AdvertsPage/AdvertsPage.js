import React from 'react';
import { Link } from 'react-router-dom';
import { Empty, Button, Spin, List, Divider } from 'antd';
import { connect } from 'react-redux';
import T from 'prop-types';

import Layout from '../../layout';
import FiltersForm, { defaultFilters } from './FiltersForm';
import AdvertCard from './AdvertCard';
import {
  getFilters,
  getAdverts,
  searchAdverts,
  apiLoading,
  getErrorMessage,
} from '../../../store/ducks/app';

export class AdvertsPage extends React.Component {
  renderLoading = () => (
    <div style={{ display: 'flex', justifyContent: 'center' }}>
      <Spin size="large" />
    </div>
  );

  renderError = () => {
    const { error } = this.props;
    return (
      <Empty
        description={<span style={{ color: '#ff4d4f' }}>{`${error}`}</span>}
      >
        <Button type="primary" danger onClick={this.loadAdverts}>
          Reload
        </Button>
      </Empty>
    );
  };

  renderEmpty = () => {
    const { filters } = this.props;
    const isFiltered =
      JSON.stringify(filters) !== JSON.stringify(defaultFilters);
    return (
      <Empty description={<span>No adverts here!</span>}>
        {isFiltered ? (
          <span>Refine your search</span>
        ) : (
          <Link to="/adverts/new">
            <Button type="primary">Create the first one</Button>
          </Link>
        )}
      </Empty>
    );
  };

  renderAdvert = advert => {
    return (
      <List.Item>
        <Link to={`/adverts/${advert._id}`}>
          <AdvertCard {...advert} />
        </Link>
      </List.Item>
    );
  };

  renderAdverts = () => {
    const { adverts, loading, error } = this.props;
    if (loading) {
      return this.renderLoading();
    }

    if (error) {
      return this.renderError();
    }

    if (!adverts) {
      return null;
    }

    if (!adverts.length) {
      return this.renderEmpty();
    }

    return (
      <List
        grid={{ gutter: 16, column: 3 }}
        dataSource={adverts}
        renderItem={this.renderAdvert}
      />
    );
  };

  loadAdverts() {
    const { filters, searchAdverts } = this.props;
    searchAdverts(filters, true);
  }

  componentDidMount() {
    this.loadAdverts();
  }

  render() {
    return (
      <Layout title="Adverts list">
        <Divider>Filter your adverts</Divider>
        <FiltersForm />
        <Divider>Adverts</Divider>
        {this.renderAdverts()}
      </Layout>
    );
  }
}

AdvertsPage.propTypes = {
  filters: T.shape({
    name: T.string,
    sale: T.string,
    price: T.arrayOf(T.number),
    tags: T.arrayOf(T.string),
  }),
  adverts: T.array.isRequired,
  loading: T.bool.isRequired,
  error: T.string,
  searchAdverts: T.func.isRequired,
};

const mapStateToProps = state => {
  return {
    filters: getFilters(state),
    adverts: getAdverts(state),
    loading: apiLoading(state),
    error: getErrorMessage(state),
  };
};

export default connect(mapStateToProps, { searchAdverts })(AdvertsPage);
