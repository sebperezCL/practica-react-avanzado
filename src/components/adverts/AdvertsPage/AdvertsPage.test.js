import React from 'react';
import { shallow } from 'enzyme';

import { AdvertsPage } from './AdvertsPage';

describe('AdvertsPage', () => {
  const props = {
    loading: false,
    error: null,
    filters: {
      name: 'name',
      sale: 'sale',
      price: [],
      tags: [],
    },
    adverts: [],
    searchAdverts: jest.fn(),
  };

  const render = () => shallow(<AdvertsPage {...props} />);

  test('snapshot', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  test('should render empty AdvertsPage with Refine your search message', () => {
    const wrapper = render();
    expect(props.searchAdverts).toHaveBeenCalled();
    expect(wrapper.find('span').text()).toBe('Refine your search');
  });

  test('should render AdvertsPage with List items', () => {
    props.adverts = [{ _id: 'id' }];
    const wrapper = render();
    expect(props.searchAdverts).toHaveBeenCalled();
    expect(wrapper.find('List').exists()).toBe(true);
  });
});
