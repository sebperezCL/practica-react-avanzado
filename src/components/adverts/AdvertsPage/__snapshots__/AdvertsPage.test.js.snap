// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvertsPage snapshot 1`] = `
<Layout
  title="Adverts list"
>
  <Divider>
    Filter your adverts
  </Divider>
  <Connect(FiltersForm) />
  <Divider>
    Adverts
  </Divider>
  <Empty
    description={
      <span>
        No adverts here!
      </span>
    }
  >
    <span>
      Refine your search
    </span>
  </Empty>
</Layout>
`;
